Battery status with channel communication

ela-kotulska-frequenz opened the pull request.

ela-kotulska-frequenz committed:
Rename BatteryPoolStatus to BatteriesStatus

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Rename _batteries_pool_status.py to _batteries_status.py

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Rename StatusTracker to BatteryStatusTracker

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Remove circular dependency in BatteriesStatus

And simplify BatteriesStatus code.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Clean BatteryStatusTracker code and improve documentation

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Move blocking feature of BatteryStatusTracker into separate class

In the future commit blocking will be only in BatteryStatusTracker.
All blocking related public functions will be removed.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Clean BatteriesStatusTracker code

Move getting status functionality to separate method.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Change tracking batteries status feature to use channel communication

BatteryStatusTracker receives messages from components, parse it.
If status change BatteryStatusTracker sends new status using channel.
BatteriesStatus is now only class that listen for change of
status from any batteries and updates set of working batteries.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

shsms commented on src/frequenz/sdk/actor/power_distributing/_batteries_status.py:
```diff
@@ -101,29 +98,25 @@ def get_working_batteries(self, battery_ids: Set[int]) -> Set[int]:
-    def update_last_request_status(self, result: Result):
+    def update_request_response(
+        self, succeed_batteries: Set[int], failed_batteries: Set[int]
+    ):
```
because this is not connected to a `Result` object anymore,  it can be renamed.  Possibly to `update_status`, `update_battery_status`, etc?  This is already inside `AllBatteryStatus`,  so maybe just `_update_from_ids` might also work.  Also, does this need to be a public method?

ela-kotulska-frequenz replied:
Done. I decided on : `update_status`

shsms replied:
does this have to be a public method?

ela-kotulska-frequenz replied:
Yes, because PowerDistributingActor actor uses it.

ela-kotulska-frequenz commented:
Unit tests are waiting for: https://github.com/frequenz-floss/frequenz-channels-python/pull/68

shsms commented on src/frequenz/sdk/microgrid/_battery/_status.py:
```diff
@@ -33,17 +33,20 @@
     NOT_WORKING = 0
-    # Component should work, but it failed in last request. It is blocked for few
-    # seconds and it is not recommended to use it until it is necessary.
+    """Component it is not working and should not be used"""
```
``` suggestion
    """Component is not working and should not be used."""
```

ela-kotulska-frequenz replied:
Done

shsms commented on src/frequenz/sdk/microgrid/_battery/_status.py:
```diff
@@ -33,17 +33,20 @@
-    # Component is working
+    """Component should work, but it failed in last request. It is blocked for few
+    seconds and it is not recommended to use it until it is necessary.
+    """
```
``` suggestion
    """Component should work, but it failed in last request. It is blocked for few
    seconds and it is not recommended to use it unless it is necessary.
    """
```

ela-kotulska-frequenz replied:
Done

shsms commented on src/frequenz/sdk/microgrid/_battery/_status.py:
```diff
@@ -45,10 +40,25 @@ class BatteryStatus(Enum):
     component_id: int
-    receiver: Peekable[T]
+    last_msg_timestamp: datetime = datetime.now(tz=timezone.utc)
+    last_msg_correct: bool = False
```
I guess you can now rename this class to `_ComponentStreamStatus`?  Also it is not generic anymore.

ela-kotulska-frequenz replied:
Done

shsms reviewed (commented):
I've checked the first 4 commits later (and one thing with the big commit).

I'll continue tomorrow.

shsms commented on src/frequenz/sdk/actor/power_distributing/_batteries_status.py:
```diff
@@ -6,33 +6,76 @@
+        working = self.working & batteries
+        if len(working) > 0:
+            return working
+        return self.uncertain & batteries
```
I guess you are already changing these into explicit function calls to `intersection`

ela-kotulska-frequenz replied:
Done

shsms commented on src/frequenz/sdk/actor/power_distributing/_batteries_status.py:
```diff
@@ -6,33 +6,76 @@
+        receiver_name = f"{self.name}-receiver"
+        self.receiver = channel.new_receiver(name=receiver_name, maxsize=1)
+
+        self.sender = channel.new_sender()
```
Based on how this is getting used, I think this should be a function instead,  with the signature:

``` python
def _create_battery_status_channel(
    battery_id: int,
) -> tuple[Sender[BatteryStatus], Receiver[BatteryStatus]]:
```

Having a class to own receivers that are accessed from somewhere else is a pattern that can be misused.  We should try to make sure that `Receiver` objects are always locally owned,  and not through a proxy, but directly.

Also,  I think it is best to avoid naming classes as `*Channel` or `*Receiver`, etc. in the sdk, unless they are actual channels or receivers deriving from the channels package.

ela-kotulska-frequenz replied:
I changed it to dataclass, renamed to _BatteryStatusChannelHelper, and I described that the channels has only 1 receiver.

It is just internal, helper class. And it is very simple, so I decided to leave it as class.

Returning tuple doesn't work, because I also need channel name as string. So I would return Tuple with 3 elements and that is a lot.
```
def _create_battery_status_channel(
    battery_id: int,
) -> tuple[Sender[BatteryStatus], Receiver[BatteryStatus]]:
```

shsms replied:
I'd say it would be more readable if this gets inlined,  but I guess it is just a matter of style.

shsms commented on src/frequenz/sdk/microgrid/_battery/_status.py:
```diff
@@ -45,10 +40,25 @@ class BatteryStatus(Enum):
 
 
+@dataclass
+class RequestResult:
```
Does this have to go to `microgrid/_battery`?  I think this is specific to the power distributor.

shsms replied:
can we also rename this to `SetPowerResult` or something?  `RequestResult` sounds too confusing.

ela-kotulska-frequenz replied:
I moved whole file to frequez.sdk.actor.power_distributing.
Renamed RequestResult to SetPowerResult.

shsms commented on src/frequenz/sdk/microgrid/_battery/_status.py:
```diff
@@ -113,12 +123,10 @@ def is_blocked(self) -> bool:
 
 
-class BatteryStatusTracker(AsyncConstructible):
+class BatteryStatusTracker:
```
This too,  I think should sit outside the `microgrid` package.  The `microgrid` package is supposed to be a thin wrapper providing access to data from the microgrid and for sending commands to the microgrid.

But this class is really stateful tracking logic, which should ideally be part of a higher level package,  possibly the `actor/power_distributing`.   I see that this code was already merged,  so I can move out later also, if you prefer.

ela-kotulska-frequenz replied:
Moved to `actor/power_distributing`

shsms commented on src/frequenz/sdk/microgrid/_battery/_status.py:
```diff
@@ -113,12 +123,10 @@ def is_blocked(self) -> bool:
-    To create an instance of this class you should use `async_new` class method.
-    Standard constructor (__init__) is not supported and using it will raise
-    `NotSyncConstructible` error.
+    Status in send only when it change.
```
``` suggestion
    Status updates are sent out only when there is a status change.
```

ela-kotulska-frequenz replied:
Done

shsms commented on src/frequenz/sdk/microgrid/_battery/_status.py:
```diff
@@ -137,18 +145,17 @@ class BatteryStatusTracker(AsyncConstructible):
-        cls, battery_id: int, max_data_age_sec: float, max_blocking_duration_sec: float
-    ) -> BatteryStatusTracker:
+    # Instance attributes defined outside __init__
+    _select: Optional[Select]
```
You don't need this line.  You can just do this inside the `__init__` method, like you're already doing for other variables:

``` python
self._select: Optional[Select] = None
```

ela-kotulska-frequenz replied:
Done

shsms commented on src/frequenz/sdk/actor/power_distributing/_batteries_status.py:
```diff
@@ -6,33 +6,76 @@
+        self.sender = channel.new_sender()
+
+
+class BatteriesStatus:
```
I guess you're going to rename this to `AllBatteryStatus` as discussed, and the file accordingly.

ela-kotulska-frequenz replied:
I renamed to `BatteryPoolStatus` once again. 
Because it has battery_ids as argument, so it can be created for pool, too.
Also `BatteryPoolStatus` sounds better after all.

shsms commented on src/frequenz/sdk/actor/power_distributing/_batteries_status.py:
```diff
@@ -46,77 +89,98 @@ async def async_new(
+
+        self._batteries: Dict[str, BatteryStatusTracker] = {}
+
+        # Receivers are needed to create Select object
```
``` suggestion
        # Receivers for individual battery statuses are needed to create a `MergeNamed` object
```

ela-kotulska-frequenz replied:
Done

shsms commented on src/frequenz/sdk/actor/power_distributing/_batteries_status.py:
```diff
@@ -46,77 +89,98 @@ async def async_new(
+            try:
+                await self._update_status(self._battery_status_channel)
+            except Exception as err:  # pylint: disable=broad-except
+                _logger.error("BatteriesStatus failed with error: %s. Restart.", err)
```
``` suggestion
                _logger.error("BatteriesStatus failed with error: %s. Restarting.", err)

ela-kotulska-frequenz replied:
Done

shsms commented on src/frequenz/sdk/actor/power_distributing/_batteries_status.py:
```diff
@@ -46,77 +89,98 @@ async def async_new(
 
-        Raises:
-            KeyError: If any battery in the given batteries is not in the pool.
+    async def _update_status(self, status_channel: MergeNamed[BatteryStatus]) -> None:
```
maybe more readable to decompose to a `Receiver`, like below,  but it is ok to keep it as it is also.
``` suggestion
    async def _update_status(self, status_channel: Receiver[tuple[str, BatteryStatus]]) -> None:
```

ela-kotulska-frequenz replied:
I would prefer to leave it `MergeNamed`.
It is clear and shorter.

shsms replied:
btw,  why not just use `self._battery_status_channel` and not pass anything at all?

shsms commented on src/frequenz/sdk/microgrid/_battery/_status.py:
```diff
@@ -197,54 +201,120 @@ def battery_id(self) -> int:
-        """Return status of the battery.
+    async def stop(self) -> None:
+        """Stop tracking battery status."""
+        if self._select is not None:
```
you should probably also `cancel_and_await(self._task)` here, before killing the `Select`.

ela-kotulska-frequenz replied:
Done

ela-kotulska-frequenz committed:
Update frequenz-channel-python to v0.13.0

... and stop MergeNamed object in BatteriesStatus.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Move BatteryStatusTracker file

Move from frequenz/sdk/microgrid/_battery/_status.py to
frequenz/sdk/actor/power_distributing/_battery_status.py

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Rename BatteriesStatus file

Rename from batteries_status.py to battery_pool_status.py

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Rename BatteriesStatus to BatteryPoolStatus

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Rename some internal classes

... battery_status.py and battery_pool_status.py

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Switch class names in battery_pool_status and battery_status

BatteryStatus will be used outside, by users.
Status is just an internal class.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>
